src/Nethermind/Nethermind.Trie/PatriciaTree.cs

```diff
@@ -138,64 +138,60 @@ public void Commit(long blockNumber, bool skipRoot = false, WriteFlags writeFlag
                 ThrowReadOnlyTrieException();
             }
 
-            if (RootRef is not null && RootRef.IsDirty)
+            int maxLevelForConcurrentCommit = -1;
+            _writeBeforeCommit /= 64;
+            if (_writeBeforeCommit > 0)
             {
-                Commit(new NodeCommitInfo(RootRef, TreePath.Empty), skipSelf: skipRoot);
-                while (TryDequeueCommit(out NodeCommitInfo node))
+                maxLevelForConcurrentCommit++; // Ok, we separate at top level
+                if (_writeBeforeCommit / 16 > 0)
                 {
-                    if (_logger.IsTrace) Trace(blockNumber, node);
-                    TrieStore.CommitNode(blockNumber, node, writeFlags: writeFlags);
+                    maxLevelForConcurrentCommit++; // Another level
                 }
-
-                // reset objects
-                TreePath path = TreePath.Empty;
-                RootRef!.ResolveKey(TrieStore, ref path, true, bufferPool: _bufferPool);
-                SetRootHash(RootRef.Keccak!, true);
             }
```

Comment:
This is equivalent of:

```
            int maxLevelForConcurrentCommit = _writeBeforeCommit switch
            {
                > 64 * 16 => 1, // we separate at two top levels
                > 64 => 0, // we separate at top level
                _ => -1
            };
```

Sounds a bit too low for me?
Did you want more something like:

```
            int maxLevelForConcurrentCommit = _writeBeforeCommit switch
            {
                > 16 * 16 * 16 => 2, // we separate at three top levels
                > 16 * 16 => 1, // we separate at two top levels
                > 16 => 0, // we separate at top level
                _ => -1
            };
```

Again, regarding unbalanced tries and unbalanced updates, would it be better to have a more dynamic task/path split based on available threads instead of a static path split?

But this would make sense to change only if this is a potential bottleneck. Do you have some profiling sessions?

Reply:
This is per tree and does not go to storage, so its fairly balanced beacuse of address/slot->hash translation. Most storage tree are have little change though, something arount 50. If I'm not mistaken, multithreading does not help there. So this mainly help the last state tree. See yellow line in the graph. Parallizing between storage need higher level interface change.

Reply:
Yeah you are probably right parallelizing storage tries (and changing the interface) is way more important.